### PR TITLE
Auth

### LINKED ISSUE #22: Login API
https://github.com/codeday-labs/bookworms/projects/1#card-65218978
~~https://github.com/codeday-labs/bookworms/projects/1#card-65219014~~

It is recommended to use **JWT** standard as it is reliable and easy to implement.

### PR DESCRIPTION
**Pull request by veritem**
This API should fix #22 

 - [x] siginin and signup
 - [x] Authorization
 - [x] Updated create review api
 - [x] Delete review api
 - [x] Like and unlike api

### CODE REVIEW

**Comment by jeromewu** [comment]
@veritem Usually we don't review a PR in `Draft` status, can you remove the `Draft` first?

File: server/docs/api.md [hunk 1]
```diff
@@ -292,3 +294,66 @@ $ curl -H "Accept: application/json" https://bookworms-api.vercel.app/api/catego
...
+### POST /api/signup
+
+
+**Example Request**
+
+```
+$ curl --location --request POST 'https://bookworms-api.vercel.app/api/signup' \
+--header 'Content-Type: application/json' \
+--data-raw '{
+    "names":"test",
```
Comment (jeromewu) [inline server/docs/api.md]: Use `name` instead as it is not an array.

File: server/utils/jwt.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	token, err := jwt.ParseWithClaims(tokenValue, claims, func(t *jwt.Token) (interface{}, error) {
+		return Jwt_key, nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if !token.Valid {
+		return nil, err
```
Comment (jeromewu) [inline server/utils/jwt.go]: err will almost be nil in this case, better to return an error with meaningful information

File: server/utils/jwt.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	"github.com/codeday-labs/bookworms/server/db"
+	"github.com/golang-jwt/jwt"
+)
+
+type Claims struct {
+	User db.User
+	jwt.StandardClaims
+}
+
+var Jwt_key = []byte("kdksksm2039")
```
Comment (jeromewu) [inline server/utils/jwt.go]: It is not a good practice to hard code jwt secret in your source, I have created an environment variable in Vercel called `JWT_SECRET` with the same value instead.

File: server/db/user.go [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+var UserCollection = "user"
+
+type User struct {
+	ID        primitive.ObjectID `bson:"_id" json:"id"`
+	Email     string             `bson:"email" json:"email"`
+	Password  string             `bson:"password" json:"password"`
+	Names     string             `bson:"names" json:"names"`
```
Comment (jeromewu) [inline server/db/user.go]: Should be singular.

File: server/api/signin.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+	}
+
+	if !utils.CheckHashPassword(request.Password, user.Password) {
+		return nil, errors.New("Email or password missmatch!"), nil
+	}
+
+	claims := &utils.Claims{
+		User: user,
+		StandardClaims: jwt.StandardClaims{
+			ExpiresAt: time.Now().Add(time.Hour * 24).Unix(),
```
Comment (jeromewu) [inline server/api/signin.go:56]: How long the token can be used should be highlighted in the document.

File: server/api/signin.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+type signinRequest struct {
+	Email    string `json:"email"`
+	Password string `json:"password"`
+}
+
+type signinResponse struct {
+	Message bool `json:"message"`
+}
+
+func signin(request signinRequest) (*signinResponse, error, *string) {
```
Comment (jeromewu) [inline server/api/signin.go]: Usually `err` is the last returned value.

File: server/api/signin.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+		if err != nil {
+			utils.RespondWithError(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		if token != nil {
+			http.SetCookie(w, &http.Cookie{
+				Name:     "token",
+				Value:    *token,
+				Expires:  time.Now().Add(time.Hour * 1),
```
Comment (jeromewu) [inline server/api/signin.go]: This part is tricky as the token is valid for 24 hours, but cookie only valid for 1 hour, better to sync there two values.

**Review by jeromewu — CHANGES_REQUESTED** [review]
I assume this PR is still far from completion as all api endpoints require update (ex. validate token) afterward.

File: server/utils/jwt.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+	"github.com/codeday-labs/bookworms/server/db"
+	"github.com/golang-jwt/jwt"
+)
+
+type Claims struct {
+	User db.User
+	jwt.StandardClaims
+}
+
+var Jwt_key = []byte(os.Getenv("kdksksm2039"))
```
Comment (jeromewu) [inline server/utils/jwt.go]: Get secret from `JWT_SECRET` instead.

```go
os.Getenv("JWT_SECRET")
```

File: server/db/review.go [hunk 9]
```diff
@@ -9,12 +9,12 @@ import (
...
-	ID         primitive.ObjectID `bson:"_id,omitempty" json:"id"`
-	UserName   string             `bson:"user_name,omitempty" json:"user_name"`
-	BookName   string             `bson:"book_name,omitempty" json:"book_name"`
-	Text       string             `bson:"text,omitempty" json:"text"`
-	Categories []string           `bson:"categories,omitempty" json:"categories"`
-	CreatedAt  time.Time          `bson:"created_at,omitempty" json:"created_at"`
-	UpdatedAt  time.Time          `bson:"updated_at,omitempty" json:"updated_at"`
-	Likes      int                `bson:"likes,omitempty" json:"likes"`
+	ID         primitive.ObjectID   `bson:"_id,omitempty" json:"id"`
+	User       primitive.ObjectID   `bson:"user,omitempty" json:"user"`
```
Comment (jeromewu) [inline server/db/review.go]: Better to use `UserID` here.

File: server/api/reviews.go [hunk 10]
```diff
@@ -253,6 +263,66 @@ func ReviewsHandler(w http.ResponseWriter, r *http.Request) {
 
 		utils.RespondWithSuccess(w, http.StatusCreated, review)
 
+	case http.MethodDelete:
```
Comment (jeromewu) [inline server/api/reviews.go:262]: Delete is not in CORS allowlist.

File: server/api/reviews.go [hunk 11]
```diff
@@ -253,6 +263,66 @@ func ReviewsHandler(w http.ResponseWriter, r *http.Request) {
...
+
+		if reviewFromID.User != user.ID {
+			utils.RespondWithError(w, "You are not allowed to delete review you don't own!", http.StatusNonAuthoritativeInfo)
+			return
+		}
+
+		_, err = DB.Collection(db.ReviewsCollection).DeleteOne(db.Ctx, bson.D{{Key: "_id", Value: objectId}})
+
+		if err != nil {
+			utils.RespondWithError(w, "Delete failed!", http.StatusInternalServerError)
```
Comment (jeromewu) [inline server/api/reviews.go:318]: Should return after error handling.

File: server/utils/jwt.go [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+func GetUserFromCookie(r *http.Request) (*db.User, error) {
```
Comment (jeromewu) [inline server/utils/jwt.go:19]: Looks like token expiration case is not covered.

File: server/db/review.go [hunk 13]
```diff
@@ -9,12 +9,12 @@ import (
...
+	BookName   string               `bson:"book_name,omitempty" json:"book_name"`
+	Text       string               `bson:"text,omitempty" json:"text"`
+	Categories []string             `bson:"categories,omitempty" json:"categories"`
+	CreatedAt  time.Time            `bson:"created_at,omitempty" json:"created_at"`
+	UpdatedAt  time.Time            `bson:"updated_at,omitempty" json:"updated_at"`
+	Likes      []primitive.ObjectID `bson:"likes,omitempty" json:"likes"`
```
Comment (jeromewu) [inline server/db/review.go:19]: As schema of review is updated, `api.md` should be updated as well.

**Review by jeromewu — CHANGES_REQUESTED** [review]
Few minor comments.

File: server/utils/jwt.go (hunk 12 above)
Comment (veritem) [reply server/utils/jwt.go:19]: Validation for that is done by the library internally, you can check line 30 
I have also done a test for that where i set up a to last for 30 seconds and it threw an error

Comment (jeromewu) [reply server/utils/jwt.go:19]: Got it, thanks!

File: server/api/reviews.go (hunk 10 above)
Comment (veritem) [reply server/api/reviews.go:262]: It's there https://github.com/veritem/bookworms/blob/f377a20c9fa0695e544ab43cf05893b4bd62bcac/server/api/reviews.go#L192

**Comment by jeromewu** [comment]
Merged, thanks for the hard work. :smile:

**Pull request merged**